src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs

```diff
@@ -1638,5 +1638,219 @@ class Source2
+        if (o is N.G<B>[]) System.Console.WriteLine(""N.G<B>[]"");
+        if ((o, o) is (A[], A[])) System.Console.WriteLine(""Twice."");
+    }
+}
+class A { }
+class G<T> { }
+class B { }
+namespace N
+{
+    class G<T> { }
+}
+";
+            var expectedOutput =
+@"int[]
+int[]
+long[]
+long[]
+A[]
+A[]
+Twice.
+G<B>[]
+G<B>[]
+N.G<B>[]
+N.G<B>[]";
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Preview), options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics(
+                );
+            var comp = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+        }
+
+        [Fact, WorkItem(40149, "https://github.com/dotnet/roslyn/issues/40149")]
+        public void ArrayTypePattern_02()
+        {
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        M1(new int[0]);
+        M1(new long[0]);
+        M1(new A[0]);
+        M1(new G<B>[0]);
+        M1(new N.G<B>[0]);
+    }
+    static void M1(object o)
+    {
+        switch (o)
+        {
+            case (int[]): System.Console.WriteLine(""int[]""); break;
+            case (System.Int64[]): System.Console.WriteLine(""long[]""); break;
+            case (A[]): System.Console.WriteLine(""A[]""); break;
+            case (G<B>[]): System.Console.WriteLine(""G<B>[]""); break;
+            case (N.G<B>[]): System.Console.WriteLine(""N.G<B>[]""); break;
+        }
+        if (o is (int[])) System.Console.WriteLine(""int[]"");
+        if (o is (System.Int32[])) System.Console.WriteLine(""long[]"");
+        if (o is (A[])) System.Console.WriteLine(""A[]"");
+        if (o is (G<B>[])) System.Console.WriteLine(""G<B>[]"");
+        if (o is (N.G<B>[])) System.Console.WriteLine(""N.G<B>[]"");
+        if ((o, o) is ((A[]), (A[]))) System.Console.WriteLine(""Twice."");
+    }
+}
+class A { }
+class G<T> { }
+class B { }
+namespace N
+{
+    class G<T> { }
+}
+";
+            var expectedOutput =
+@"int[]
+int[]
+long[]
+long[]
+A[]
+A[]
+Twice.
+G<B>[]
+G<B>[]
+N.G<B>[]
+N.G<B>[]";
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Preview), options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics(
+                );
+            var comp = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+        }
+
+        [Fact, WorkItem(40149, "https://github.com/dotnet/roslyn/issues/40149")]
+        public void ParsedAsExpressionBoundAsType_01()
+        {
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        M1(new A());
+        M1(new G<B>.D());
+        M1(new N.G<B>.D());
+        M1(0);
+    }
+    static void M1(object o)
+    {
+        switch (o)
+        {
+            case A: System.Console.WriteLine(""A""); break;
+            case G<B>.D: System.Console.WriteLine(""G<B>.D""); break;
```

Comment:
>G<B>.D [](start = 17, length = 6)

Is there a semantic model API we could use to probe these labels?

Reply:
Not supported by this PR.

---
In reply to: [362990971](https://github.com/dotnet/roslyn/pull/40503#discussion_r362990971) [](ancestors = 362990971)